core/src/Streamly/Internal/Data/Unbox/TH.hs

```diff
@@ -0,0 +1,412 @@
+                ]
+
+    where
+
+    lenCons = length cons
+    tagType = getTagType lenCons
+    tagSize = getTagSize lenCons
+    peekMatch (i, con) =
+        match (litP (IntegerL i)) (normalB (mkPeekExprOne tagSize con)) []
+    peekErr =
+        match
+            wildP
+            (normalB
+                 [|error
+                       ("Found invalid tag while peeking (" ++
+                        $(lift (pprint headTy)) ++ ")")|])
+            []
+
+--------------------------------------------------------------------------------
+-- Poke
+--------------------------------------------------------------------------------
+
+mkPokeExprTag :: Name -> Int -> Q Exp
+mkPokeExprTag tagType tagVal = pokeTag
+
+    where
+
+    pokeTag =
+        [|pokeByteIndex
+              $(varE _initialOffset)
+              $(varE _arr)
+              $((sigE (litE (IntegerL (fromIntegral tagVal))) (conT tagType)))|]
+
+mkPokeExprFields :: Int -> [Field] -> Q Exp
+mkPokeExprFields tagSize fields = do
+    case fields of
+        [] -> [|pure ()|]
+        _ ->
+            letE
+                (mkOffsetDecls tagSize fields)
+                (doE $ map (noBindS . pokeField) [0 .. (numFields - 1)])
+
+    where
+
+    numFields = length fields
+    pokeField i =
+        [|pokeByteIndex
+              $(varE (mkOffsetName i))
+              $(varE _arr)
+              $(varE (mkFieldName i))|]
+
+mkPokeMatch :: Name -> Int -> Q Exp -> Q Match
+mkPokeMatch cname numFields exp0 =
+    match
+        (conP cname (map varP (map mkFieldName [0 .. (numFields - 1)])))
+        (normalB exp0)
+        []
+
+mkPokeExpr :: Type -> [DataCon] -> Q Exp
+mkPokeExpr headTy cons =
+    case cons of
+        [] ->
+            [|error
+                  ("Attempting to poke type with no constructors (" ++
+                   $(lift (pprint headTy)) ++ ")")|]
+        [(DataCon _ _ _ [])] -> [|pure ()|]
+        [(DataCon cname _ _ fields)] ->
+            caseE
+                (varE _val)
+                [mkPokeMatch cname (length fields) (mkPokeExprFields 0 fields)]
+        _ ->
+            caseE
+                (varE _val)
+                (map (\(tagVal, (DataCon cname _ _ fields)) ->
+                          mkPokeMatch
+                              cname
+                              (length fields)
+                              (doE [ noBindS $ mkPokeExprTag tagType tagVal
+                                   , noBindS $ mkPokeExprFields tagSize fields
+                                   ]))
+                     (zip [0 ..] cons))
+
+    where
+
+    lenCons = length cons
+    tagType = getTagType lenCons
+    tagSize = getTagSize lenCons
+
+
+--------------------------------------------------------------------------------
+-- Main
+--------------------------------------------------------------------------------
+
+-- XXX Document this function properly
+makeUnboxWith :: Cxt -> Type -> [DataCon] -> Q [Dec]
```

Comment:
Write doc, what additional functionality does it provide compared to Unbox? Does it need to be released?

Reply:
We need not release it.